flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java

```diff
@@ -315,8 +326,20 @@ public FutureCompletingBlockingQueue<RecordsWithSplitIds<E>> getQueue() {
      * @throws Exception when failed to close the split fetcher manager.
      */
     public synchronized void close(long timeoutMs) throws Exception {
+        final long startTime = System.currentTimeMillis();
         closed = true;
         fetchers.values().forEach(SplitFetcher::shutdown);
+        // Actively drain the element queue in case there are previously shutting down
+        // fetcher threads blocking on putting batches into the element queue.
+        executors.submit(
+                () -> {
+                    while (fetchersToShutDown.get() > 0
+                            && System.currentTimeMillis() - startTime < timeoutMs) {
+                        elementsQueue
+                                .getAvailabilityFuture()
+                                .thenRun(() -> elementsQueue.poll().recycle());
+                    }
+                });
         executors.shutdown();
         if (!executors.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)) {
             LOG.warn(
```

Comment:
(I can't add my comment on the exact line due to the limit of the GitHub webpage) I think the warning log here should print `fetchersToShutDown` instead of `fetchers.size()`.